runtime/include/cloe/data_broker.hpp

```diff
@@ -0,0 +1,2063 @@
+/*
+ * Copyright 2023 Robert Bosch GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+/**
+ * \file cloe/data_broker.hpp
+ *
+ * Classes:
+ *   DataBroker:
+ *     A central registry for (type erased) signals.
+ *
+ *   Signal:
+ *     A type-erased abstraction of a signal.
+ *
+ *   BasicContainer<T>
+ *     An optional container storing the value of a signal and abstracting the interaction with Signal.
+ *
+ * Background:
+ *   Real world simulations utilize manifolds of signals (variables) and coresponding types.
+ *   Still the very basic problem around a single signal (variable) boils down to CRUD
+ *   - declaring a signal (Create)
+ *   - reading its value / receiving value-changed events (Read)
+ *   - writing its value / triggering value-changed events (Update)
+ *   - NOT NEEDED: undeclaring a signal (Delete)
+ *
+ * Concept:
+ *   Abstract CRU(D) operations by registering signals in a uniform way by their name.
+ */
+
+#pragma once
+#ifndef CLOE_DATA_BROKER_HPP_
+#define CLOE_DATA_BROKER_HPP_
+
+#include <any>
+#include <cassert>
+#include <functional>
+#include <map>
+#include <memory>
+#include <regex>
+#include <stdexcept>
+#include <string_view>
+#include <typeindex>
+#include <vector>
+
+#include <fmt/format.h>
+
+#include <sol/sol.hpp>
+
+namespace cloe {
+
+namespace databroker {
+
+namespace detail {
+
+/**
+  * Detects the presence of the to_lua function (based on ADL)
+  */
+template <typename T, typename = void>
+struct has_to_lua : std::false_type {};
+/**
+  * Detects the presence of the to_lua function (based on ADL)
+  */
+template <typename T>
+struct has_to_lua<
+    T, std::void_t<decltype(to_lua(std::declval<sol::state_view>(), std::declval<T*>()))>>
+    : std::true_type {};
+/**
+  * Detects the presence of the to_lua function (based on ADL)
+  */
+template <typename T>
+constexpr bool has_to_lua_v = has_to_lua<T>::value;
+
+/**
+  * Invokes to_lua procedure, if detecting its presence
+  */
+template <typename T>
+void to_lua(sol::state_view lua) {
+  if constexpr (has_to_lua_v<T>) {
+    to_lua(lua, static_cast<T*>(nullptr));
```

Comment:
SEVERITY_MEDIUM_HIGH What? :D I know you need that parameter T* functionwise but I from a users perspective this does not make sense. People implementing the to_lua function will ask what they are supposed to do with that parameter.

(As per my understanding, the to_lua function is supposed to be implemented as a member function of T)

Reply:
And having a look at the implementation of the to_lua function in the Wheel class: According to my own trials, the data type of usertype_table["rotation"] is double Wheel::* - I dont understand what exactly is happening tbh. Could you explain this to me once?